cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SourceControlManagementService.java

```diff
@@ -169,4 +192,56 @@ public PushAppResponse pushApp(ApplicationReference appRef, @Nullable String com
 
     return pushResponse;
   }
+
+  /**
+   * Pull the application from linked repository and deploy it in current namespace.
+   * @param appRef application reference to deploy with
+   * @return {@link ApplicationRecord} of the deployed application.
+   * @throws Exception when {@link ApplicationLifecycleService} fails to deploy.
+   */
+  public ApplicationRecord pullAndDeploy(ApplicationReference appRef) throws Exception {
+    PullAppResponse pullResponse = validatePulledApplication(appRef);
+
+    AppRequest<?> appRequest = DECODE_GSON.fromJson(pullResponse.getAppDetailString(), AppRequest.class);
+    KerberosPrincipalId ownerPrincipalId =
+      appRequest.getOwnerPrincipal() == null ? null : new KerberosPrincipalId(appRequest.getOwnerPrincipal());
+
+    // if we don't null check, it gets serialized to "null"
+    Object config = appRequest.getConfig();
+    String configString = config == null ? null : config instanceof String ? (String) config : GSON.toJson(config);
+    ChangeSummary changeSummary = appRequest.getChange();
+    SourceControlMeta sourceControlMeta = new SourceControlMeta(pullResponse.getFileHash());
+
+    // Deploy with a generated uuid
+    String versionId = RunIds.generate().getId();
+    ApplicationWithPrograms app = appLifecycleService.deployApp(
+      appRef.getParent(), appRef.getApplication(), versionId, appRequest.getArtifact(), configString,
+      changeSummary, sourceControlMeta, x -> { }, ownerPrincipalId, appRequest.canUpdateSchedules(),
+      false, Collections.emptyMap()
+    );
+
+    LOG.info("Successfully deployed app {} in namespace {} from artifact {} with configuration {} and " +
+               "principal {}", app.getApplicationId().getApplication(), app.getApplicationId().getNamespace(),
+             app.getArtifactId(), configString, app.getOwnerPrincipal()
+    );
+
+    return new ApplicationRecord(
+      ArtifactSummary.from(app.getArtifactId().toApiArtifactId()),
+      app.getApplicationId().getApplication(),
+      app.getApplicationId().getVersion(),
+      app.getSpecification().getDescription(),
+      Optional.ofNullable(app.getOwnerPrincipal()).map(KerberosPrincipalId::getPrincipal).orElse(null),
+      app.getChangeDetail(), app.getSourceControlMeta());
+  }
+
+  private PullAppResponse validatePulledApplication(ApplicationReference appRef) throws NoChangesToPullException {
+    SourceControlMeta latestMeta = store.getAppSourceControlMeta(appRef);
+    PullAppResponse pullResponse = sourceControlOperationRunner.pull(appRef);
+
+    if (latestMeta != null && latestMeta.getFileHash().equals(pullResponse.getFileHash())) {
```

Comment:
file hash can be null in certain edge cases like the one we discussed in the other PR

Reply:
Added null check and made it nullable in `SourceControlMeta`